airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/InitialSyncCtidIterator.java

```diff
@@ -303,13 +303,12 @@ public PreparedStatement createCtidLegacyQueryStatement(final Connection connect
     Preconditions.checkArgument(lowerBound != null, "Lower bound ctid expected");
     Preconditions.checkArgument(upperBound != null, "Upper bound ctid expected");
     try {
-      LOGGER.info("*** one more {}", lowerBound);
       LOGGER.info("Preparing query for table: {}", tableName);
       final String fullTableName = getFullyQualifiedTableNameWithQuoting(schemaName, tableName,
           quoteString);
       final String wrappedColumnNames = RelationalDbQueryUtils.enquoteIdentifierList(columnNames, quoteString);
       final String sql =
-          "SELECT ctid::text, %s FROM %s WHERE ctid = ANY (ARRAY (SELECT FORMAT('(%%s,%%s)', page, tuple)::tid FROM generate_series(?, ?) as page, generate_series(?,?) as tuple))"
+          "SELECT ctid::text, %s FROM %s WHERE ctid = ANY (ARRAY (SELECT FORMAT('(%%s,%%s)', page, tuple)::tid tid_addr FROM generate_series(?, ?) as page, generate_series(?,?) as tuple ORDER BY tid_addr))"
```

Comment:
Do we need to check for any other queries that bring in values we use in checkpointing that also need `order by`?

Reply:
the other ctid query is `WHERE ctid > '0,0' AND ctid <= '(131000,0)'` 
this >< range is returning rows in sequential order - it was created as an optimization for old postgres. So I don't think we need to sort there ourselves.

With the postgres 12 query, there is some logic applied by postgres that it iterates over the bigger of two ranges first (pages) . I couldn't find a way to control it other than sorting the array. 

Other incremental queries are all or nothing - so if an error happened in the middle of incremental xmin for example, it will not checkpoint at all. so even if there is some case records are out of order we will be good (I don't think this can happen anywhere) .

Reply:
My general recommendation is to always add a sort when you want data sorted. 

SQL query optimizers can easily remove the sorting step when it's not needed (or chose a query plan that's cheaper with an implicit sort than one that would be cheaper without a sort but for which the sort is expensive). 
Query plans can also change from release to release, which could change the order of the results, absent an explicit sort. The real signal to look for is a change in query cost and query performance.

Reply:
I agree Stephane 
I added an explicit test that is taking a large amount of records (> single page) and makes sure records are received in order.
Because the TID Range Scan for newer postgres versions was [created with the purpose](https://www.postgresql.org/message-id/E1lFwPH-0002fZ-QI%40gemulon.postgresql.org) of making an optimized scan, I'd prefer to not add an expensive sort. there can be millions and millions of records on each chunk we read

Reply:
@rodireich but if pg 14+ is already sorting when issuing a `WHERE ctid > '0,0' AND ctid <= '(131000,0)'` scan I agree with @stephane-airbyte that we should add an explicit ORDER BY here. 

There are 2 scenarios : 
1. It's already ordering by ctid in which case the query optimizer should ignore this. 
2. It isn't doing already doing this and the ORDER BY query adds significant latency.

If the order by for PG14+ adds significant latency, we can always : 
1. Tune chunk size so that the ORDER BY occurs in memory.
2. Checkpoint at the end of every chunk and keep track of the largest CTID entry (and avoid the ORDER BY query at all)

I'm worried about a similar case in PG14+ where there are some cases where hte TID range scan is not returning records in order.